### ISSUE TITLE
"endTime only valid when runState is terminated" on version 1.4.6

### ISSUE THREAD

**Issue by mrballcb** [thelastpickle/cassandra-reaper#749]
We are running cassandra 3.11.4 and reaper 1.4.6 in kubernetes clusters and hitting a symptom that seems to have hit a couple people here so far.

Before today, we were running reaper 1.2.2 and the repairs would consistently fail (as well as manual repairs).  We upgraded reaper to 1.4.6 and it began acting like it was going to work because it detected that it needed to upgrade something (guessing the migration scripts updating the repair_run table).  But as we continued to watch it, it restarted itself multiple times, and we realized it was complaining that it couldn't "resume running repair runs".

```
ERROR  [2019-09-09 21:10:22,259] [ReaperApplication-scheduler] i.c.ReaperApplication - Couldn't resume running repair runs
io.cassandrareaper.ReaperException: java.lang.IllegalStateException: endTime only valid when runState is terminated. RUNNING 2019-09-09T19:29:30.451Z
  at io.cassandrareaper.service.RepairManager.resumeRunningRepairRuns(RepairManager.java:112)
  at io.cassandrareaper.ReaperApplication.lambda$scheduleRepairManager$0(ReaperApplication.java:293)
  at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.IllegalStateException: endTime only valid when runState is terminated. RUNNING 2019-09-09T19:29:30.451Z
  at com.google.common.base.Preconditions.checkState(Preconditions.java:721)
  at io.cassandrareaper.core.RepairRun$Builder.build(RepairRun.java:303)
  at io.cassandrareaper.storage.CassandraStorage.buildRepairRunFromRow(CassandraStorage.java:1217)
  at io.cassandrareaper.storage.CassandraStorage.getRepairRunsWithStateForCluster(CassandraStorage.java:724)
  at io.cassandrareaper.storage.CassandraStorage.getRepairRunsWithState(CassandraStorage.java:704)
  at io.cassandrareaper.service.RepairManager.resumeRunningRepairRuns(RepairManager.java:105)
  ... 9 common frames omitted
ERROR  [2019-09-09 21:10:24,067] [ReaperApplication-scheduler] i.c.ReaperApplication - Failed purging repair runs from history
java.lang.IllegalStateException: endTime only valid when runState is terminated. RUNNING 2019-09-09T19:29:30.451Z
  at com.google.common.base.Preconditions.checkState(Preconditions.java:721)
  at io.cassandrareaper.core.RepairRun$Builder.build(RepairRun.java:303)
  at io.cassandrareaper.storage.CassandraStorage.buildRepairRunFromRow(CassandraStorage.java:1217)
  at io.cassandrareaper.storage.CassandraStorage.getRepairRunsAsync(CassandraStorage.java:685)
  at io.cassandrareaper.storage.CassandraStorage.getRepairRunsForCluster(CassandraStorage.java:657)
  at io.cassandrareaper.service.PurgeService.purgeDatabase(PurgeService.java:63)
  at io.cassandrareaper.ReaperApplication.lambda$schedulePurge$1(ReaperApplication.java:312)
  at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241)
  at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
  at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
  at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
  at java.lang.Thread.run(Thread.java:748)
```

The new version appears to be happy with the schema it finds when it starts up:
```
INFO   [2019-09-09 21:10:08,576] [main] o.c.c.m.MigrationRepository - Found 19 migration scripts
INFO   [2019-09-09 21:10:08,577] [main] i.c.s.CassandraStorage - Keyspace reaper_db already at schema version 19
```

Drawing hints from the the two previous issues that were opened, I took a look at the number of rows in the `reaper_db.repair_run` table with `state` values:
ABORTED: 0
ERROR: 3418
PAUSED: 0
DONE: (I dunno, I hit the 10 second timeout in the cqlsh, not really sure how to find out this number)
RUNNING: 6239  (of which 5064 have end_time = null)

That last one seems like the one it's complaining about, right?  It's saying that the 1175 running repairs (impossible I believe) that have the state "RUNNING" already have an end_time associated with them, which is invalid data as far as reaper is concerned.  Correct?

Should reaper fix this?  Or is there some dangerous "reaper shouldn't touch this" condition I can't conceive of?  (Or am I way off base with this supposition?)

I'm happy to provide any debugging info.  I'm not a cassandra expert though (far from it), so I might need a little assistance coming up with answers to things you would like to see.  I have cqlsh access, and reaper cli access, and reaper UI access and can provide just about anything you request as long as I understand what it is you want.

**Comment by mrballcb** [comment]
I also wanted to mention that we have 5 cassandra DCs around the world all replicating to each other (low volume).  Whatever I do to that `reaper_db.repair_run` table, I need to make sure I'm only doing it for *this* cluster and not one of the other 4.

**Comment by adejanovski** [comment]
Hi @mrballcb,

you can fix that issue by setting `endTime` to `null` on all runs that have a `RUNNING` state.
`endTime`  should only be set when the repair run is in a terminated state (`DONE`, `ERROR`, `ABORTED` or `DELETED`).
I'm not sure how your DB got corrupted this way, but we had a few issues with this on 1.2.x versions indeed.

Reaper should start smoothly once the database is fixed.

**Comment by mrballcb** [comment]
We scaled down the reaper deployments on all of our clusters.  Then any record which was still `RUNNING` that had an end_time, I changed the state field to `ABORTED`.  When I started up the reapers again, they did begin to start doing their normal work.  I've been watching the number of rows in the `reaper_db.repair_run` table with state=`RUNNING` but end_time=null, and it's been slowly going down.  Thus it seems to be properly processing (resuming?) and setting the state to `DONE` or something.  It looks like it's working correctly once again (to me).

Thank you so much for the guidance.
